### PR TITLE
*: add Cancel method for tidb.Session, Canceled for context.Context

### PR DESCRIPTION
**Pull request by tiancaiamao**
* add a Cancel method to Session  interface
* add a Canceled method to Context interface
* introduce a ctxForCancel struct, which would be used for canceling ticlient later

To release resource safety, we should take special care for the following places :
1. distsql goroutine in executor
2. join goroutine in executor
3. transaction in store/tikv
4. distsql & ticlient 

This PR introduce some method and struct, and handle part of 1, https://github.com/pingcap/tidb/pull/1988 already support part of 3,
2 and 4 will be added in the following PR.

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -915,3 +925,32 @@ func statementContextToFlags(sc *variable.StatementContext) uint64 {
...
+		// NOTE: Not thread safe, but it should work.
+		ctx.ch = make(chan struct{})
+		close(ctx.ch)
+	}
+	// Receive from a nil channel blocks forever.
+	return ctx.ch
+}
+
+func (ctx *ctxForCancel) Deadline() (deadline time.Time, ok bool) {
+	ok = false
```
Comment (zimulala) [inline executor/distsql.go]: The default value of `ok` is false, remove this line.

File: executor/distsql.go [hunk 2]
```diff
@@ -915,3 +925,32 @@ func statementContextToFlags(sc *variable.StatementContext) uint64 {
...
 	return flags
 }
+
+// ctxForCancel implements the standard Go context, provide Done() method.
+type ctxForCancel struct {
+	ctx context.Context
+	ch  chan struct{}
+}
+
+func (ctx *ctxForCancel) Done() <-chan struct{} {
```
Comment (zimulala) [inline executor/distsql.go]: Can you move this method to the `session`?

**Comment by tiancaiamao** [comment]
PTAL @zimulala

File: distsql/distsql_test.go [hunk 3]
```diff
@@ -15,6 +15,7 @@ package distsql
 
 import (
 	"bytes"
+	"context"
```
Comment (coocood) [inline distsql/distsql_test.go]: alias to `goctx` to avoid confusion.

File: distsql/distsql.go [hunk 4]
```diff
@@ -14,6 +14,7 @@
 package distsql
 
 import (
+	"context"
```
Comment (coocood) [inline distsql/distsql.go]: alias to goctx to avoid confusion.

File: session.go [hunk 5]
```diff
@@ -108,6 +115,18 @@ type session struct {
 	sessionVars *variable.SessionVars
 }
 
+// Cancel cancels the execution of current transaction.
+func (s *session) Cancel() {
+	// TODO: how to wait for the resource to release and make sure
```
Comment (zimulala) [inline session.go]: s/how/How

File: session.go [hunk 6]
```diff
@@ -93,8 +96,12 @@ func (h *stmtHistory) add(stmtID uint32, st ast.Statement, params ...interface{}
 }
 
 type session struct {
-	txn    kv.Transaction // current transaction
-	txnCh  chan *txnWithErr
+	txn   kv.Transaction // current transaction
+	txnCh chan *txnWithErr
+	// For cancel the execution of current transaction.
+	txnCtx        goctx.Context
```
Comment (coocood) [inline session.go]: rename to `goCtx` is better.

File: session.go [hunk 7]
```diff
@@ -93,8 +96,12 @@ func (h *stmtHistory) add(stmtID uint32, st ast.Statement, params ...interface{}
...
+	txnCancelFunc goctx.CancelFunc
```
Comment (coocood) [inline session.go]: rename to `cancelFunc` is better.

**Pull request merged**